SteaneReedMuller Code

Fe-r-oz opened the pull request.

Fe-r-oz commented:
```
improved SRM(1,r)

Outputs:

julia> parity_checks(SteaneReedMuller(1,3))
+ XXXXXXXX
+ ZZZZZZZZ
+ __ZZXXYY
+ _ZXY_ZXY
+ _XZYZY_X

julia> parity_checks(SteaneReedMuller(1,4))
+ XXXXXXXXXXXXXXXX
+ ZZZZZZZZZZZZZZZZ
+ ____ZZZZXXXXYYYY
+ __ZZXXYY__ZZXXYY
+ _ZXY_ZXY_ZXY_ZXY
+ _X_XZYZYZYZY_X_X

julia> parity_checks(SteaneReedMuller(1,5))
+ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+ ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ
+ ________ZZZZZZZZXXXXXXXXYYYYYYYY
+ ____ZZZZXXXXYYYY____ZZZZXXXXYYYY
+ __ZZXXYY__ZZXXYY__ZZXXYY__ZZXXYY
+ _ZXY_ZXY_ZXY_ZXY_ZXY_ZXY_ZXY_ZXY
+ _X_X_X_XZYZYZYZYZYZYZYZY_X_X_X_X
canonicalize:

julia> canonicalize!(parity_checks(SteaneReedMuller(1,3)))
+ XZ_YZXY_
+ _XZYZY_X
+ _ZXY_ZXY
+ __ZZXXYY
+ ZZZZZZZZ

julia> canonicalize!(parity_checks(Gottesman(3)))
+ XZ_Y_YXZ
+ _XZY_XZY
- _ZXYZ_YX
- __ZZYYXX
+ ZZZZZZZZ
```

had to close #249 because of the merge conflicts introduced by @250.

Gottesman subset is being tests with official Gottesman test.

Fe-r-oz commented:
It was told that the Stabilizer can accept two bool arguments, no need to pass ;UINt8 argument, but the following happens when testing for large stabilizer circuits;

Stabilizer(extended_Hx, extended_Hz) 
`
 TypeError: in typeassert, expected Matrix{UInt64}, got a value of type Vector{Any}`

SteaneReedMuller(1,3) to (1,7) don't have this error.

Krastanov commented:
> It was told that the Stabilizer can accept two bool arguments, no need to pass ;UINt8 argument, but the following happens when testing for large stabilizer circuits;
> 
> Stabilizer(extended_Hx, extended_Hz) ` TypeError: in typeassert, expected Matrix{UInt64}, got a value of type Vector{Any}`
> 
> SteaneReedMuller(1,3) to (1,7) don't have this error.

An easy way to check what constructors are available is:

```
julia> methods(Stabilizer)
# 10 methods for type constructor:
  [1] Stabilizer(s::Stabilizer)
  [2] Stabilizer(xzs::AbstractMatrix{Bool})
  [3] Stabilizer(phases::Tzv, nqubits::Int64, xzs::Tm) where {Tzv<:AbstractVector{UInt8}, Tm<:(AbstractMatrix{<:Unsigned})}
  [4] Stabilizer(paulis::AbstractArray{PauliOperator{Tz, Tv}, 1}) where {Tz, Tv}
  [5] Stabilizer(phases::AbstractVector{UInt8}, xs::AbstractMatrix{Bool}, zs::AbstractMatrix{Bool})
  [6] Stabilizer(phases::AbstractVector{UInt8}, xzs::AbstractMatrix{Bool})
  [7] Stabilizer(xs::AbstractMatrix{Bool}, zs::AbstractMatrix{Bool})
  [8] Stabilizer(tab::T) where T<:QuantumClifford.Tableau
  [9] Stabilizer(g::Graphs.SimpleGraphs.SimpleGraph)
 [10] Stabilizer(c::QuantumClifford.ECC.AbstractECC)
```

I think you have been using method 5, but method 7 should be more convenient. Given your error message, it seems the problem is that somewhere a `Vector{Any}` is being fed instead of a `Matrix{Bool}`. More information is necessary to pinpoint the exact issue.

Also, it is worrisome that this happens only for some sizes. It implies there might be some size-dependent type-instability, which is usually a bad sign.

Fe-r-oz commented:
The Hx and Hz are both Boolean Matrices before they are passed down to the Stabilizer.

I will check out how the vector{any} comes into play.  Also,  I will introspect on the tip: size-dependent type-instability,

Krastanov commented:
On the master branch tests pass. You can disregard nightly/performance/downgrade/buildkite, but the rest should pass here. Let's keep this in draft form until the tests pass.

Krastanov commented:
To run the tests locally, you can do `] test QuantumClifford`. Avoid pushing here until things work locally.

Fe-r-oz commented:
```
julia> naive_encoding_circuit(parity_checks(SteaneReedMuller(3,6)))
690-element Vector{QuantumClifford.AbstractOperation}:
 QuantumClifford.sHadamard(1)
 QuantumClifford.sZCX(1,43)
 QuantumClifford.sZCX(1,44)
 QuantumClifford.sZCX(1,45)
 QuantumClifford.sZCX(1,46)
 QuantumClifford.sZCX(1,47)
 QuantumClifford.sZCX(1,49)
 QuantumClifford.sZCX(1,50)
 QuantumClifford.sZCX(1,51)
 QuantumClifford.sZCX(1,52)
 QuantumClifford.sZCX(1,54)
 QuantumClifford.sZCX(1,55)
 QuantumClifford.sZCX(1,56)
 QuantumClifford.sZCX(1,58)
 QuantumClifford.sZCX(1,59)
 QuantumClifford.sZCX(1,61)
 QuantumClifford.sHadamard(2)
 ⋮
 QuantumClifford.sSWAP(33,27)
 QuantumClifford.sSWAP(32,28)
 QuantumClifford.sSWAP(31,29)
 QuantumClifford.sSWAP(30,28)
 QuantumClifford.sSWAP(29,26)
 QuantumClifford.sSWAP(28,26)
 QuantumClifford.sSWAP(27,25)
 QuantumClifford.sSWAP(26,24)
 QuantumClifford.sSWAP(25,23)
 QuantumClifford.sSWAP(24,23)
 QuantumClifford.sSWAP(23,22)
 QuantumClifford.sSWAP(22,21)
 QuantumClifford.sSWAP(21,20)
 QuantumClifford.sSWAP(20,19)
 QuantumClifford.sSWAP(19,18)
 QuantumClifford.sSWAP(18,17)
 QuantumClifford.sSWAP(17,16)

julia> naive_encoding_circuit(parity_checks(SteaneReedMuller(3,4)))
127-element Vector{QuantumClifford.AbstractOperation}:
 QuantumClifford.sHadamard(1)
 QuantumClifford.sPhase(1)
 QuantumClifford.sZCX(1,16)
 QuantumClifford.sHadamard(2)
 QuantumClifford.sZCX(2,16)
 QuantumClifford.sHadamard(3)
 QuantumClifford.sZCZ(3,2)
 QuantumClifford.sZCX(3,16)
 QuantumClifford.sHadamard(4)
 QuantumClifford.sZCX(4,16)
 QuantumClifford.sHadamard(5)
 QuantumClifford.sPhase(5)
 QuantumClifford.sZCZ(5,3)
 QuantumClifford.sZCX(5,16)
 QuantumClifford.sHadamard(6)
 QuantumClifford.sPhase(6)
 QuantumClifford.sZCZ(6,4)
 ⋮
 QuantumClifford.sZCZ(15,8)
 QuantumClifford.sZCZ(15,9)
 QuantumClifford.sZCZ(15,10)
 QuantumClifford.sZCZ(15,11)
 QuantumClifford.sZCZ(15,12)
 QuantumClifford.sZCZ(15,13)
 QuantumClifford.sZCX(15,16)
 QuantumClifford.sZ(1)
 QuantumClifford.sZ(2)
 QuantumClifford.sZ(3)
 QuantumClifford.sZ(4)
 QuantumClifford.sZ(5)
 QuantumClifford.sZ(6)
 QuantumClifford.sZ(7)
 QuantumClifford.sZ(9)
 QuantumClifford.sZ(11)
 QuantumClifford.sZ(13)
```
```

julia> naive_encoding_circuit(parity_checks(SteaneReedMuller(5,6)))
1732-element Vector{QuantumClifford.AbstractOperation}:
 QuantumClifford.sHadamard(1)
 QuantumClifford.sPhase(1)
 QuantumClifford.sZCX(1,64)
 QuantumClifford.sHadamard(2)
 QuantumClifford.sZCX(2,64)
 QuantumClifford.sHadamard(3)
 QuantumClifford.sZCZ(3,2)
 QuantumClifford.sZCX(3,64)
 QuantumClifford.sHadamard(4)
 QuantumClifford.sZCX(4,64)
 QuantumClifford.sHadamard(5)
 QuantumClifford.sZCZ(5,3)
 QuantumClifford.sZCX(5,64)
 QuantumClifford.sHadamard(6)
 QuantumClifford.sZCZ(6,4)
 QuantumClifford.sZCX(6,64)
 QuantumClifford.sHadamard(7)
 ⋮
 QuantumClifford.sZ(36)
 QuantumClifford.sZ(37)
 QuantumClifford.sZ(38)
 QuantumClifford.sZ(39)
 QuantumClifford.sZ(41)
 QuantumClifford.sZ(42)
 QuantumClifford.sZ(44)
 QuantumClifford.sZ(46)
 QuantumClifford.sZ(47)
 QuantumClifford.sZ(49)
 QuantumClifford.sZ(50)
 QuantumClifford.sZ(51)
 QuantumClifford.sZ(54)
 QuantumClifford.sZ(55)
 QuantumClifford.sZ(57)
 QuantumClifford.sZ(59)
 QuantumClifford.sZ(61)
```

Fe-r-oz commented:
code_k(c::....) = ....will give 0 at (4, 2) based on Eq 2. Steane calls it ''code size falls to 0.'' Not sure how the base test perceives k =0 for frame or other calculations that require k.

I've come to appreciate that even my mistakes uncover loopholes in the tests. The base test is more comprehensive now as a result.

Fe-r-oz commented:
I don't know precisely the definition of  code_s(c::...) so I didn't included it. Maybe it's total number of syndromes bits?

I have included `code_k, code_n. ` `Distance` is 2^c.t + 2^(c.t-1) as the improved base requires them for calculations of `frames`

If there are improvements that can be made, if further attention is required in tests, etc. please let me know.

Fe-r-oz commented:
Maybe it's better to include a sentence in docstring why Generator matrics are actually stabilizers meaning same as parity check tableaus for some values, this could really prevent confusion if people see parity check Tableau and ask "Why is formed in the structure of Generator Matrixs which is G = H as follows

```
 [ G1| 0 ]  
 [ 0| G2 ]
 [ Dx|Dz ]

```

`G = H (in some cases)
`
After reading the paper, the reason is that 

_"The method works for k > 0, and the generators for k < 0 are stabilisers of k > 0 codes."_


It would be helpful if permission is granted to add this line in the docstring.

Fe-r-oz commented:
>I don't know precisely the definition of code_s(c::...) so I didn't included it. Maybe it's total number of syndromes bits?

As per def. of `code_s` in `CSS`:
`code_s(c::CSS) = size(c.Hx, 1) + size(c.Hz, 1)`

I guess code_s is total rows of stabilizer since in CSS, we have different rows for Hz and Hx. This can be seen in different examples of CSS.

In Steane's case, I have something like

`code_s(c::...) = _nplusk(c.t, c.r) `

since that's tells us total number of rows in this case.

Fe-r-oz commented:
> More information is necessary to pinpoint the exact issue.
> [5] Stabilizer(phases::AbstractVector{UInt8}, xs::AbstractMatrix{Bool}, zs::AbstractMatrix{Bool})
>[7] Stabilizer(xs::AbstractMatrix{Bool}, zs::AbstractMatrix{Bool})

Very sorry for not providing complete details. My bad. 

Problem was in the when` naive_encoding_circuit `(which starts from line 18 to line 53), Maybe because in [5], we provide a parameter phases, (what is phased btw?) and in [7], we don't, so maybe it expects a unit8 phase parameter?

Complete appropriate trace:
I have added (please see here in the trace). This is from where we see the source in`naive_encoding_circuit`
```
 BoundsError: attempt to access 16-element view(::Vector{UInt8}, 17:32) with eltype UInt8 at index [17]
  Stacktrace:
    [1] throw_boundserror(A::SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}, I::Tuple{Int64})
      @ Base ./abstractarray.jl:737
    [2] checkbounds
      @ ./abstractarray.jl:702 [inlined]
    [3] getindex
      @ ./subarray.jl:322 [inlined]
    [4] naive_encoding_circuit(code::QuantumClifford.ECC.SteaneReedMuller; undoperm::Bool)
      @ QuantumClifford.ECC ~/Desktop/New/b10/QuantumClifford.jl/src/ecc/circuits.jl:53 (please see here)
    [5] naive_encoding_circuit
      @ ~/Desktop/New/b10/QuantumClifford.jl/src/ecc/circuits.jl:18 [inlined] (please see here)
    [6] test_with_pframes(code::QuantumClifford.ECC.SteaneReedMuller)
      @ Main.var"##ecc#256" ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc.jl:47
```

Krastanov commented:
It might help to run the other ECC tests as well, to see what fails. The one about encoding&syndrome circuits is probably too involved, with too many pieces that work together, so it is difficult to debug.

What other tests fail? In particular, what are the `test_ecc_codeproperties` results.

It is indeed quite probably that the issue is the k=0 codes. If it happens only for them, we should: (1) understand what specifically about the assumptions in the tests is wrong and (2) filter out these codes only for these specific tests.

Fe-r-oz commented:
Breakdown after running each encoding and decoding file separately 

Firstly, 
`
test_ecc_decoder_all_setups` pass for the given values of SteaneReedMuller
```
julia> include("test_ecc_decoder_all_setups.jl")
WARNING: redefinition of constant Main.code_instance_args. This may fail, cause incorrect answers, or produce other errors.
Test Summary:                       |Time
table decoder, good for small codes | None  8.1s
Test Summary:                               |Time
belief prop decoders, good for sparse codes | None  0.1s
Test Summary:                                                          |Time
matching decoder, good as long as column weight of the code is limited | None  9.2s
Test.DefaultTestSet("matching decoder, good as long as column weight of the code is limited", Any[], 0, false, false, true, 1.712044735925158e9, 1.712044745110257e9, false, "/home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_decoder_all_setups.jl")
```

K<0 error in test_ecc.

```
julia> include("test_ecc.jl")
WARNING: redefinition of constant Main.code_instance_args. This may fail, cause incorrect answers, or produce other errors.
naive syndrome circuits - zero syndrome for logical states: Error During Test at /home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc.jl:37
  Got exception outside of a @test
  ArgumentError: dimension size must be ≥ 0, got -10 for dimension 1
  Stacktrace:
    [1] BitArray
      @ ./bitarray.jl:32 [inlined]
    [2] BitArray
      @ ./bitarray.jl:71 [inlined]
    [3] falses(dims::Tuple{Int64})
      @ Base ./bitarray.jl:405
    [4] falses
      @ ./bitarray.jl:403 [inlined]
    [5] quantum_mallows(rng::Random._GLOBAL_RNG, n::Int64)
      @ QuantumClifford ~/Desktop/New/b10/QuantumClifford.jl/src/randoms.jl:189
    [6] random_destabilizer(rng::Random._GLOBAL_RNG, n::Int64; phases::Bool)
      @ QuantumClifford ~/Desktop/New/b10/QuantumClifford.jl/src/randoms.jl:73
    [7] random_destabilizer
      @ ~/Desktop/New/b10/QuantumClifford.jl/src/randoms.jl:72 [inlined]
    [8] random_stabilizer(rng::Random._GLOBAL_RNG, n::Int64; phases::Bool)
      @ QuantumClifford ~/Desktop/New/b10/QuantumClifford.jl/src/randoms.jl:166
    [9] random_stabilizer
      @ ~/Desktop/New/b10/QuantumClifford.jl/src/randoms.jl:166 [inlined]
   [10] random_stabilizer
      @ ~/Desktop/New/b10/QuantumClifford.jl/src/randoms.jl:167 [inlined]
   [11] test_naive_syndrome(c::SteaneReedMuller, e::Bool)
      @ Main ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc.jl:12
   [12] macro expansion
      @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc.jl:39 [inlined]
   [13] macro expansion
      @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
   [14] top-level scope
      @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc.jl:38
   [15] include(fname::String)
      @ Base.MainInclude ./client.jl:489
   [16] top-level scope
      @ REPL[7]:1
   [17] eval
      @ ./boot.jl:385 [inlined]
   [18] eval_user_input(ast::Any, backend::REPL.REPLBackend, mod::Module)
      @ REPL ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:150
   [19] repl_backend_loop(backend::REPL.REPLBackend, get_module::Function)
      @ REPL ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:246
   [20] start_repl_backend(backend::REPL.REPLBackend, consumer::Any; get_module::Function)
      @ REPL ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:231
   [21] run_repl(repl::REPL.AbstractREPL, consumer::Any; backend_on_current_task::Bool, backend::Any)
      @ REPL ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:389
   [22] run_repl(repl::REPL.AbstractREPL, consumer::Any)
      @ REPL ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:375
   [23] (::Base.var"#1013#1015"{Bool, Bool, Bool})(REPL::Module)
      @ Base ./client.jl:432
   [24] #invokelatest#2
      @ ./essentials.jl:892 [inlined]
   [25] invokelatest
      @ ./essentials.jl:889 [inlined]
   [26] run_main_repl(interactive::Bool, quiet::Bool, banner::Bool, history_file::Bool, color_set::Bool)
      @ Base ./client.jl:416
   [27] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:333
   [28] _start()
      @ Base ./client.jl:552
Test Summary:                                              | Pass  Error  Total  Time
naive syndrome circuits - zero syndrome for logical states |  320      1    321  2.5s
ERROR: LoadError: Some tests did not pass: 320 passed, 0 failed, 1 errored, 0 broken.
in expression starting at /home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc.jl:37


```

test_ecc_codeproperties: I have not provided the code_s in the file as I don't know it's definition. 
```

julia> include("test_ecc_codeproperties.jl")
WARNING: redefinition of constant Main.code_instance_args. This may fail, cause incorrect answers, or produce other errors.
Test Summary: | Pass  Total  Time
is CSS        |   21     21  0.4s
code tableau consistency: Test Failed at /home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:34
  Expression: code_s(code) + code_k(code) == code_n(code)
   Evaluated: 6 == 16

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] macro expansion
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:34 [inlined]
 [3] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [4] top-level scope
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:30
code tableau consistency: Test Failed at /home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:35
  Expression: size(H, 1) < size(H, 2)
   Evaluated: 16 < 16

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] macro expansion
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:35 [inlined]
 [3] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [4] top-level scope
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:30
code tableau consistency: Test Failed at /home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:38
  Expression: QuantumClifford.stab_looks_good(copy(H))

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] macro expansion
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:38 [inlined]
 [3] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [4] top-level scope
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:30
code tableau consistency: Test Failed at /home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:35
  Expression: size(H, 1) < size(H, 2)
   Evaluated: 64 < 64

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] macro expansion
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:35 [inlined]
 [3] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [4] top-level scope
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:30
code tableau consistency: Test Failed at /home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:38
  Expression: QuantumClifford.stab_looks_good(copy(H))

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] macro expansion
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:38 [inlined]
 [3] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [4] top-level scope
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:30
code tableau consistency: Test Failed at /home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:34
  Expression: code_s(code) + code_k(code) == code_n(code)
   Evaluated: 8 == 64

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] macro expansion
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:34 [inlined]
 [3] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [4] top-level scope
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:30
code tableau consistency: Test Failed at /home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:35
  Expression: size(H, 1) < size(H, 2)
   Evaluated: 64 < 64

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] macro expansion
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:35 [inlined]
 [3] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [4] top-level scope
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:30
code tableau consistency: Test Failed at /home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:38
  Expression: QuantumClifford.stab_looks_good(copy(H))

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] macro expansion
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:38 [inlined]
 [3] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [4] top-level scope
   @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:30
Test Summary:            | Pass  Fail  Total  Time
code tableau consistency |  118     8    126  0.7s
ERROR: LoadError: Some tests did not pass: 118 passed, 8 failed, 0 errored, 0 broken.
in expression starting at /home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_codeproperties.jl:29

```

syndromes

```
include("test_ecc_syndromes.jl")
WARNING: redefinition of constant Main.code_instance_args. This may fail, cause incorrect answers, or produce other errors.
naive and shor measurement circuits: Error During Test at /home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_syndromes.jl:45
  Got exception outside of a @test
  BoundsError: attempt to access 16-element view(::Vector{UInt8}, 17:32) with eltype UInt8 at index [17]
  Stacktrace:
    [1] throw_boundserror(A::SubArray{UInt8, 1, Vector{UInt8}, Tuple{UnitRange{Int64}}, true}, I::Tuple{Int64})
      @ Base ./abstractarray.jl:737
    [2] checkbounds
      @ ./abstractarray.jl:702 [inlined]
    [3] getindex
      @ ./subarray.jl:322 [inlined]
    [4] naive_encoding_circuit(code::SteaneReedMuller; undoperm::Bool)
      @ QuantumClifford.ECC ~/Desktop/New/b10/QuantumClifford.jl/src/ecc/circuits.jl:53
    [5] naive_encoding_circuit
      @ ~/Desktop/New/b10/QuantumClifford.jl/src/ecc/circuits.jl:18 [inlined]
    [6] pframe_naive_vs_shor_syndrome(code::SteaneReedMuller)
      @ Main ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_syndromes.jl:10
    [7] macro expansion
      @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_syndromes.jl:47 [inlined]
    [8] macro expansion
      @ ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
    [9] top-level scope
      @ ~/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_syndromes.jl:46
   [10] include(fname::String)
      @ Base.MainInclude ./client.jl:489
   [11] top-level scope
      @ REPL[5]:1
   [12] eval
      @ ./boot.jl:385 [inlined]
   [13] eval_user_input(ast::Any, backend::REPL.REPLBackend, mod::Module)
      @ REPL ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:150
   [14] repl_backend_loop(backend::REPL.REPLBackend, get_module::Function)
      @ REPL ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:246
   [15] start_repl_backend(backend::REPL.REPLBackend, consumer::Any; get_module::Function)
      @ REPL ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:231
   [16] run_repl(repl::REPL.AbstractREPL, consumer::Any; backend_on_current_task::Bool, backend::Any)
      @ REPL ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:389
   [17] run_repl(repl::REPL.AbstractREPL, consumer::Any)
      @ REPL ~/.julia/juliaup/julia-1.10.2+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:375
   [18] (::Base.var"#1013#1015"{Bool, Bool, Bool})(REPL::Module)
      @ Base ./client.jl:432
   [19] #invokelatest#2
      @ ./essentials.jl:892 [inlined]
   [20] invokelatest
      @ ./essentials.jl:889 [inlined]
   [21] run_main_repl(interactive::Bool, quiet::Bool, banner::Bool, history_file::Bool, color_set::Bool)
      @ Base ./client.jl:416
   [22] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:333
   [23] _start()
      @ Base ./client.jl:552
Test Summary:                       | Pass  Error  Total  Time
naive and shor measurement circuits |   88      1     89  0.4s
ERROR: LoadError: Some tests did not pass: 88 passed, 0 failed, 1 errored, 0 broken.
in expression starting at /home/Transcendence/Desktop/New/b10/QuantumClifford.jl/test/test_ecc_syndromes.jl:45
```

Fe-r-oz commented:
What's does `stab_looks_good `actually checks?

Krastanov commented:
> What's does `stab_looks_good `actually checks?

This list of properties https://github.com/QuantumSavory/QuantumClifford.jl/blob/11ae38c7342c76a75ad9201b609b352d38117256/src/QuantumClifford.jl#L1082

Krastanov commented:
> K<0

I think I understand what k=0 is. I do not think it makes sense to have situations where k<0.

Krastanov commented:
> Every other test seem to work well. It's seems to me that it's the k = 0, and the way the code_base is not to be setup does provides error when k = 0.
> 
> Please have a look at CI tests. They provide the same error that is locally.

It is not obvious to me that the other tests work. If I am reading this right, all these tests are not executed because the test before them failed. https://github.com/QuantumSavory/QuantumClifford.jl/blob/master/test/runtests.jl#L62-L74

Fe-r-oz commented:
> Every other test seem to work well. 

I phrased it wrong as a general statement. Testing individually, `test_ecc_decoder_all_setups.jl ` passes.

Krastanov commented:
> > I think I understand what k=0 is. I do not think it makes sense to have situations where k<0.
> 
> I have been trying to ask whether we should include k = 0 for sometime ;/ Even prepared an equation image for it!
> 
> I will add an condition that does not include k =0 cases to avoid cases which have the following error ` ArgumentError: dimension size must be ≥ 0, got -10 for dimension 1`
> 
> I will make sure that the verify with Steane Paper, so that the algorithm is correct.
> 
> Thanks for your comments! I will try to improve the work.

To clarify, k=0 can make perfect sense. I do not understand what k<0 is.

Fe-r-oz commented:
> I do not understand what k<0 is.

This  k<0 has been dangerous for me as well ;/ I don't understand how 'generators for k < 0 are stabilizers of k > 0 codes.' 

Given my limited knowledge on the subject as I am just starting out, I have not read how generator matrices are considered as stabilizers/parity matrices when k goes negative.

The paper does not elaborate on this either.

Fe-r-oz commented:
Docstring added from paper which tries to explain the  k< 0. 
I wanted to add unicode for subscript z, but I am not sure how to add that. 
These codes have peculiar properties, and thus, maybe we need to less the strictness some code properties.

Fe-r-oz commented:
Seeking out review so that I can make it better.  

>The paper does not elaborate on this either.

Paper has 1 small paragraph about it, so I had carefully added it in docstring.

Fe-r-oz commented:
Kindly please review this as well.  Looking forward to your comments!

Krastanov commented:
I will look into it sometime late next week. Ping me if I have not by then.

Fe-r-oz commented:
Removed unnessary complexity.

Removed the cases where construction method fails as per Steane.

Significantly making is simpler, easy to understand

Removed unnecessary helper function

Fe-r-oz commented:
Dear @Krastanov I am looking forward to your comments and review!

Krastanov commented on CHANGELOG.md:
```diff
@@ -5,6 +5,10 @@
 
 # News
 
+## v0.9.5 - 2024-04-14
```
There will be merge conflicts with CHANGELOG as other things are being merged. I do not have a great solution for this issue. One thing we can do is to just avoid writing in the changelog until the moment this is being merged -- if you would like that, feel free to just post what the changelog update should be in the PR description when you make a PR.

Fe-r-oz commented:
> There are a lot of code instances that are not tested for correctness. Adding these to the test runner would make it 

Thanks for your comments. I will work on the changes. Our limit is from 1 to 7 for both r and t, and most of the code instances, such as entire t =5 given n , t =4 given n have k < 0 so that will give argument error.

I will check all the cases when k > 0 such as (2,5),  (2, 6), (2 ,7), (3,7). Also, I will form larger table to see which values give k> 0 when t =6, and t =7

Fe-r-oz commented on src/ecc/codes/steanereedmuller.jl:
```diff
@@ -0,0 +1,80 @@
+    return 2^r - sum(binomial.(r, 0:t)) 
+end 
+
+function _steane_convention_qrm(mat)
```
I will remove this function. In the case, we would have to do the shifting right words instead of left words because of the changes in matrix convention.

Fe-r-oz commented on test/test_ecc_base.jl:
```diff
@@ -9,7 +9,8 @@ const code_instance_args = Dict(
     Gottesman => [3, 4, 5],
-    CSS => (c -> (parity_checks_x(c), parity_checks_z(c))).([Shor9(), Steane7(), Toric(4,4)])
+    CSS => (c -> (parity_checks_x(c), parity_checks_z(c))).([Shor9(), Steane7(), Toric(4,4)]),
+    SteaneReedMuller => [(1,3), (1,4), (1, 5), (1, 7)]
```
Yeah, I will add all the cases where k > 0. There are not many though, (2,5) , (2, 6), (2,7), and (3,7)

I will have to check whether for t =6, and t =7, which values of n provide a valid construction.

Fe-r-oz commented on src/ecc/codes/steanereedmuller.jl:
```diff
@@ -0,0 +1,80 @@
+    function SteaneReedMuller(t, r)
+        if  t < 0 || t > 7 || r < 1 || r > 7
+            throw(ArgumentError("Invalid parameters when attempting to construct a Steane-Reed-Muller code. We need  0<r≤7 and 0≤t≤7 in order to obtain a valid code and to remain tractable."))
+        elseif (2^r - binomial(r, t) - 2*sum(binomial.(r, 0:t - 1)))  < 0 || (2^r - binomial(r, t) - 2*sum(binomial.(r, 0:t - 1)))  == 0
```
Oh. That's seems much better

Fe-r-oz commented:
> reminder: please do not push after every single small edit. Try to structure your pull request and do pushes only when something substantial is ready.

Thanks for your comments. I have applied the code review suggestions, deleted the steane_convention helper function and adder proper docstring. Locally, there is problem with using `LDPC Decoders  "0.3.1" and building the gives error. The problems goes away by reverting back to 0.3.0

ERROR: Unsatisfiable requirements detected for package LDPCDecoders [3c486d74]:
 LDPCDecoders [3c486d74] log:
 ├─possible versions are: 0.1.0-0.3.0 or uninstalled
 ├─restricted to versions 0.3.1-0.3 by QuantumClifford [0525e862], leaving only versions: uninstalled

Fe-r-oz commented on src/ecc/codes/steanereedmuller.jl:
```diff
@@ -0,0 +1,80 @@
+"""The family of non-CSS Steane-Reed-Muller codes, as discovered by Steane in his 1999 paper [steane1999quantum](@cite).
+
+The construction method is effective for k > 0, and generators G  = (G_x | G_z) for k < 0 serve as stabilizers (parity check matrices) H = (H_x | H_z) for k > 0 codes.
```
I have added the explanation provided by Steane with a bit more description and also explained the parameters as well. I hope that it makes sense now.

Fe-r-oz replied:
Please let me know does docstring makes sense now? Or I should add more details? I tried to include what was provided in the Steane Paper.

Fe-r-oz commented:
The CSS `QuantumReedMuller` code has been defined. Closing this in favor of this non-CSS one.
